### DISCUSSION TITLE
Global Illumination in Filament : Present and Future

### DISCUSSION THREAD

**Discussion by Nolram12345** [google/filament#6527] [Ideas]
At the moment, Global Illumination seems to be rather absent from Filament. This is a problem for a variety of reasons, as it greatly limits applications in rendering full scenes. 
There is a todo in the projects tab for lighting about spherical harmonics global illumination (probes?). Are there any updates on this ? 

Besides this, there is also the matter of the emerging technology of raytracing. Vulkan has had its ray extensions for a while now, and we are now also getting to the point where even consumer mobile devices are getting hardware-accelerated raytracing support, so getting support for this (for instance for raytraced ambient and specular occlusion, shadows, reflections or global illumination/indirect lighting) in Filament would truly be remarkable. This has been briefly brought up before, but I don't think it has received the consideration it deserves.

Thanks !

**Comment by romainguy** [comment]
We've of course been thinking of ray-tracing but our efforts with Filament are mostly oriented towards mobile use cases and it doesn't make sense at this point for us to spend resources supporting ray-tracing.

And GI remains something we want to solve eventually (and we've discussed several ideas on how to do it), but it also hasn't been a priority over other features unfortunately.

**Comment by Nolram12345** [comment]
I understand that Filament is oriented towards mobile - however, I think that compromising on feature set for that reason may not be ideal. Filament could have a much stronger presence in multi-platform and even Desktop usage with only minimal further investment (the base PBR and GTAO are already very good). And ray-tracing is coming to mobile, even if its still in its infancy, so its more of a question of _when_ it makes sense to approach it.

As for general global illumination, I am surprised it hasn't been a bigger issue in the past. Filament has very strong rendering capabilities already, and a solid (preferably meshless) lighting pipeline could really bring it to all new heights, and I would love to see where it gets taken in that regard. Thanks for your consideration !

**Reply by romainguy** [reply]
I am not disagreeing with anything you are saying, but we have commitments to partners and users of the library and it's not what they require at this time. It's only a matter of prioritization and resources.

**Reply by Nolram12345** [reply]
I totally understand ! I just wanted to make sure intents are clear here :) Thank you for your time. Maybe sometime in the future some resources can be allocated to it.

**Reply by BStringhamVRSK** [reply]
Just adding my 2 cents in support of GI support in Filament.  We have both a desktop and mobile user base.  It's already a great engine and this would make it even better.

**Reply by BStringhamVRSK** [reply]
@romainguy 
If I may suggest one GI option worth considering: the Signed-Distance-Field (SDFGI) run-time global illumination as implemented by the Godot team.  Their solution is also open source and available for other engines to implement.  One nice feature of SDFGI is it requires no baking or preprocessing and supports both indirect lighting and reflections.
https://www.youtube.com/watch?v=QFKPrDv-Ue8

**Reply by romainguy** [reply]
We’re aware of sdfgi.

**Comment by nianbenn** [comment]
So two years have passed. Does Filament now support ray tracing?
